### DISCUSSION TITLE
feature: configurable in-combat-state during spell effects over time

### DISCUSSION THREAD

**Discussion by ValeFolken** [AscensionGameDev/Intersect-Engine#2216] [Ideas]
### Description

I would like to suggest an alteration to the Spell Menu in which you may toggle whether or not a spell puts you into combat.  This would result in the capability to apply Buffs such as stat enhancements and HP/MP Regen without the loss of out-of-combat Regen granted in the Class Editor Menu.

Here is what it looks like as of right now, unchanged
![spellcombatbefore](https://github.com/AscensionGameDev/Intersect-Engine/assets/127902267/80d44908-d50b-4698-8fc3-3118ec0cdad5)

Here is how I envision it looking once modified.  A No Combat option is there, but greyed out until you select "Friendly".  This way, you still cannot use offensive spells and then log out, cheating the system.
![spellcombatafter1](https://github.com/AscensionGameDev/Intersect-Engine/assets/127902267/4b659f41-c36e-4f62-b192-54f7e4cc720c)

Here you can see that once "Friendly" is toggled on, the capability to toggle "No Combat" becomes active.
![spellcombatafter2](https://github.com/AscensionGameDev/Intersect-Engine/assets/127902267/c5e5f117-fe00-4b3e-8183-4adfca6a9113)

And lastly, just for funsies and completionist's sake, here you see both options toggled on.
![spellcombatafter3](https://github.com/AscensionGameDev/Intersect-Engine/assets/127902267/b5fe670f-25d8-46b1-8d4b-9eb7bfff795a)

### Intended Use-Case

This enhancement would expand developers' options on making stat-enhancing and buff capabilities on items, which is a staple in many MMORPGs.  It would prevent the current issue of HP-Regen spells "overwriting" out-of-combat HP Regen, and it would prevent issues logging out when affected by very long spell durations.

Example:  My revenant has 10% HP Regen as per the Class Editor screen, thus granting him 10% health regen each server tick.  If he casts a Regen Spell on himself (Or in my case, eats a meal (item) which grants him a buff (spell) that increases HP by 10 every 10 seconds for 10 minutes) then it will put him into combat, therefore disabling his natural 10% out-of-combat regen and preventing him from safely logging out for 10 minutes.

With my suggestion implemented, he will instead be granted his natural 10% out-of-combat regen in addition to the food effects (spell) of 10 HP every 10 seconds over 10 minutes.

### Duplicate Bug Check

- [X] This enhancement idea is not a duplicate to the best of my knowledge.

**Comment by WeylonSantana** [comment]
You say "using a food", we see "uses items"
so in the same way that they use items outside of combat, they use them in combat, I don't see this as a bug, but as a behavior intended by the team. Your case would be specific and we would have to create a "mode/type" of items that do not activate the combat trigger when used? I wait for the response from the staff to say what they think

**Comment by Arufonsu** [comment]
First of all, the wording was already fixed in 7.2 beta 15, you are essentially _**casting spells with items**_.
Anyways: a spell, no matter how is structured or configured like, will by _design_ trigger the _in combat state_ for players. i don't think that this deserves the tag of a _bug_ considering that this is working as expected by design, also, FYI you can make items that heal by themselves without having to tie them to spells (which trigger the in combat state that you are trying to avoid).

Edit: i've added the *help wanted* tag and decided to not close this since it might be worth to wait for the maintainer or other collaborators to give their insight on this, i somewhat believe that we should have some sort of checkbox among the spells editor which allows it to _choose wether we want for X spell to trigger the in combat state or not._ - as for now, this is in theory and by design not an issue, but something that probably **requires an enhancement instead**.

**Comment by ValeFolken** [comment]
Oh I must have an old version then.  When I set an item to consumable, it only gives me the option to Heal, restore MP, and grant EXP.  I don't see any option for a 30 minute regen buff without combat.  I was under the impression that if I wanted an item to grant me a 30 minute Regen buff without combat, that I had to create a regen spell, and use the item to grant the spell.  Is there another way of doing this?

**Comment by Arufonsu** [comment]
> Oh I must have an old version then. When I set an item to consumable, it only gives me the option to Heal, restore MP, and grant EXP. I don't see any option for a 30 minute regen buff without combat. I was under the impression that if I wanted an item to grant me a 30 minute Regen buff without combat, that I had to create a regen spell, and use the item to grant the spell. Is there another way of doing this?

i guess that i misunderstood that part (going to change the title again 😆 ), regardless, iirc: to have such effect over time you need to use a spell, which by design send the players into combat state.

**Comment by ValeFolken** [comment]
I completely understand, that's what I have done to the letter.  However I'm unsatisfied that A 30m Regen buff working outside of combat isn't possible (Has no one in this community created this bread and butter MMO staple?) I suppose let's change this to an enhancement request then.  I see the solution going one of three ways.

1) Like you suggested, we simply have a checkbox indicating whether or not we enter combat when applying the spell.

2) Forgive my lack of syntax and context for not knowing the language, but we copy Class Editor version of HP/MP Regen into spells as an option or

3) We expand the possibilities of what a Consumable can do, such as timed buffs (including the Class Editor version of HP/MP Regen) and heck, might as well throw in Attack, Defense, etc.  And in the tooltip it would be cited as a "Buff" instead of a "Spell".

I feel I may be stepping over a line here in a community I'm unfamiliar with, but it's hard for me to imagine after all the wonderful things this engine is capable of that a Regen Food Buff that doesn't break the game's balance and logout features is out of the question, so I'll stick my neck out and ask.

**Comment by Arufonsu** [comment]
> This means that not only can players not log out

This is actually the intended behavior for every spell, this was designed by the engine's author sometime ago in order to prevent people to simply vanish out of combat in order to cheat their way out and survive 😆 which in a way, is a handy safelock over having nothing like before.

> But it prevents out-of-combat healing as well

I just took a second read on this and, can you elaborate more on this? i don't see how being in combat state doesn't allows a player to heal themselves on scenarios where they aren't really in combat and such, i'm probably missing something here o.O... regardless, yes, having a checkbox could probably fit in faster and fill this need which in a way is pretty valid but again, i don't see how the combat state wouldn't allow things to simply flow while in game with your food that heals over time, but i do see the annoying drawback where you wouldn't be able to logout until you are fully healed or the buff is over - which in your case is pretty long for a single buff... a workaround would be to use a cleanse / debuff spell item / spell, but sure, nobody wants the player to be worrying about that, the checkbox would allow it in this case to simply logout the character while cleaning up the buff which iirc, should happen within the logout method by default.

**Comment by Arufonsu** [comment]
> I suppose let's change this to an enhancement request then

Sure, it sounds more accurate to be fair 👍🏽 considering that this is, in fact, a feature that already exists, but we could improve it by adding the possibility to enable/disable it when preferred.

**Comment by ValeFolken** [comment]
I'll elaborate on the out of combat healing bit.  In the Class Editor section where you're designing the stats of the class, the exp curve, etc.  There is a HP and MP Regen % stat.  When you put 10% into HP, then you restore 10% of your max HP every server tick (frequency can be edited in server config) but only while out of combat (it doesn't explicitly state this, but it is how it functions in reality).  These Regens completely freeze whilst in combat.

**My scenario**:  A fighter class has 10% Regen in the Class Editor Menu, which shows up in his Character Sheet ingame (C).  This regen only works when completely out of combat and doesn't work at all while in combat.  Now, I have a cooking skill that allows me to make food, and I catch fish and cook a meal that applies a 30 minute Buff on me to restore 10 health every 10 seconds.  **My ideal situation** is that when I'm low on health AND out of combat, I should be receiving the Class Editor HP Regen ontop of the 10HP/10s 30 minute buff.  Instead, **what actually happens** is that the Class Editor HP Regen completely stops so long as the Food Buff is active, causing potentially an overall penalty to HP regen as a whole, depending on the ratio of time in combat vs time out of combat.

I hope this brings more detail to the situation.

**Comment by Arufonsu** [comment]
Whenever you can, please update the main post following the enhancement format which is essentially:

- _Delete_: Steps to Reproduce / Version with bug / Last version that worked well / Affected platforms / Did you find any workaround? / Relevant log output
- Update _Description_ : Please give us a detailed description of the enhancement or quality of life change you have in mind. Focus on what this change does, over the why you want it. What would be the end-result when implemented? Feel free to add any diagrams, (mocked up) screenshots, or other materials that support your vision.
- Add _Intended Use-Case_ - Describe the problem that this change would solve. Provide an example of what you are trying to achieve and how this change would help.  ie. ability to enable/disable in-combat-state for specific spells and/or their effects over time.
- Update _Duplicate Check_ - "This enhancement idea is not a duplicate to the best of my knowledge."

I've updated the title as well, i hope that this helps with your enhancement idea 👍🏽

**Comment by ValeFolken** [comment]
All done.  Let me know how that is.
